### DISCUSSION TITLE
Is there any way to get navigation pending state in layout (Next 13)?

### DISCUSSION THREAD

**Discussion by karandwivedi42** [vercel/next.js#44459] [Help]
React docs show a nice way to indicate that a transition is happening by fading the layout container (See demo here: https://beta.reactjs.org/reference/react/Suspense#indicating-that-a-transition-is-happening)

Is there any way to achieve the same in next.js during navigation?

More specifically, I want the layout to be slightly dimmed out while a router navigation is happening in its children pages.

**Comment by ruthv1k** [comment]
Hey @karandwivedi42, 

If you are using the beta version, you can make use of the new file convention for loading states (`loading.js`) to achieve this. You can read about it more here https://beta.nextjs.org/docs/routing/loading-ui

If you are using the stable or previous versions of next, You can use `useEffect`, below is a quick example (using tailwind CSS)
```tsx
function MyApp({ Component, pageProps }: AppProps) {
    const router = useRouter()

    useEffect(() => {
        const overlay = document.querySelector(".overlay")
        if (overlay) {
            overlay.classList.remove("opacity-0")
            overlay.classList.add("opacity-100")
            const timer = setTimeout(() => {
                overlay.classList.remove("opacity-100")
                overlay.classList.add("opacity-0")
            }, 500)

            return () => clearTimeout(timer);
        }
    }, [router.pathname])

    return (
        // layout
        <div className="relative">
            {/* overlay */}
            <div className="overlay opacity-0 transition-opacity duration-300 bg-black/10 w-screen h-screen absolute top-0 left-0 z-0 flex items-center justify-center">
                <LoaderIcon />
            </div>

            {/* content */}
            <div className="absolute top-0 left-0 z-10">
                <Component {...pageProps} />
            </div>
        </div>
    )
}
```

![image](https://user-images.githubusercontent.com/91245898/210133318-b2e5b3d9-6828-48dd-954e-9e2a40d5ceac.png)

You can also use a package like framer motion to perform layout animations

**Reply by karandwivedi42** [reply]
I am using the new file convention.

`loading.js` gives me a loading state for the page component but I want something to handle loading for my `layout.tsx` page.

Also, instead of changing opacity for a specific duration (500 in your case), I want to link it to the actual page transition.

`useTransition` exposes both `isPending` and `startTransition`. I want to access the `isPending` from the router transition in the `layout.js` component

**Reply by ruthv1k** [reply]
Hey @karandwivedi42, sorry for the late response

I'm not sure if there's any efficient way that I know of to achieve this in the current version, as we do not have access to `router.events` (as mentioned in the [docs](https://beta.nextjs.org/docs/api-reference/use-router)) but if you want to use something like `isPending` you might have to do it  the classic way using a `useEffect`. 

```tsx
  const currentPath = usePathname()
  const [isPending, setIsPending] = useState(true)

  useEffect(() => {
      setIsPending(false)
  }, [currentPath])
```

I'll also be keeping an eye out for any better solutions that might come up in the future.

**Reply by karandwivedi42** [reply]
Alternative discussed with a working solution [here](https://github.com/vercel/next.js/discussions/41745#discussioncomment-4127651)
